airflow/models/dagrun.py

```diff
@@ -65,6 +65,17 @@ class DagRun(Base, LoggingMixin):
         UniqueConstraint('dag_id', 'run_id'),
     )
 
+    def __init__(self, dag_id=None, run_id=None, execution_date=None, start_date=None, external_trigger=None,
+                 conf=None, state=None, **kwargs):
+        self.dag_id = dag_id
+        self.run_id = run_id
+        self.execution_date = execution_date
+        self.start_date = start_date
+        self.external_trigger = external_trigger
+        self.conf = conf
+        self.state = state
+        super().__init__(**kwargs)
```

Comment:
Is this function not what the base constructor from SQLA does? Do we really need this?

Reply:
It depends...

All SQLA model classes can be initialised with arguments via a bit of dynamic construction (specifically [this](https://docs.sqlalchemy.org/en/13/orm/extensions/declarative/api.html#sqlalchemy.ext.declarative.api._declarative_constructor) function) and as a result IDEs and Pylint don't know which arguments those classes accept (although after googling it appears Pycharm professional does have SQLAlchemy support).

Explicitly adding an `__init__` resolves this issue and allows for type hinting, autocompletion and all the other fancy stuff. So it's not absolutely necessary but could improve coder productivity. @ashb WDYT?

Reply:
It's worth noting that some DB model classes in Airflow already have an __init__, e.g. Connection, DagPickle, Log and TaskFail.

Reply:
Yeah. I am also for explicit naming of _init_ parameters (explicit is better than implicit). It makes it so much easier to understand what is going on especially when we add more type annotations as we go. It's again very similar story of making the code more readable and easy to understand by people who just want to contribute a fix or two. As a general rule - I tend to write a bit more code than bare minimum if I can say something explicitly - at the end the code will be read many times more than written and optimisation for reading "pleasure" ;) is important.

Reply:
@ashb - are you ok with that ? We could merge it now ;).

Reply:
Sure, this sounds sensible. Do we need kwargs at all then?

Reply:
Hard to say - That's the problem with dynamic constructor params. :). 

It looks like we don't need kwargs from the way how DagRun() is executed (there are just handful of places).  But @BasPH  - maybe worth checking and running the tests :).

Reply:
@potiuk not necessarily, it would allow users to set arguments on the SQLAlchemy DeclarativeMeta class but I don't see any use case for that so I'll remove them.